### PR TITLE
Maya bioin 918 compress gvcf

### PR DESCRIPTION
**Pull request by mayalevy**
cromshell_stat a7a6cbaf-cc5c-456c-b6c7-088b7a0f1202
Examples can be found here.

Also review:
https://github.com/Ultimagen/terra_pipeline/pull/481

### CODE REVIEW

File: ugvc/joint/compress_gvcf.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    ap_var.add_argument("--output_path", help="Output gvcf file path", required=True, type=str)
+    return ap_var
+
+
+def run(argv: List[str]):
+    """Makes the g.vcf file smaller by merging similar sequential records
+    The merge is done in the following way:
+    We merge sequential records as long as there GQ value of all the records in this merge is not different
+    from each other by more than 10
+    In addition, we want to keep RefCall records with GQ<22 and not merge them.
```
Comment (ilyasoifer) [inline ugvc/joint/compress_gvcf.py]: Hardcoded 22?

File: ugvc/joint/compress_gvcf.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    Returns
+    -------
+    tuple(int,int)
+        Number of records in input g.vcf, number of records in compressed g.vcf
+
+    """
+    parser = get_parser()
+    args = parser.parse_args(argv[1:])
+
+    gvcf_in = pysam.VariantFile(args.input_path, "r")
```
Comment (ilyasoifer) [inline ugvc/joint/compress_gvcf.py]: consider: with pysam.VariantFile(...) as gvcf_in

File: ugvc/joint/compress_gvcf.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    gvcf_out = pysam.VariantFile(args.output_path, "w", header=gvcf_in.header)
+    fetch_gvcf = gvcf_in.fetch()
+    sample_name = list(gvcf_in.header.samples)[0]
```
Comment (ilyasoifer) [inline ugvc/joint/compress_gvcf.py]: What happens if there are more than one sample in file? 
assert?

**Review by ilyasoifer — CHANGES_REQUESTED** [review]
PASS lines should stay

File: ugvc/joint/compress_gvcf.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                        default=10,
+                        type=int)
+    return ap_var
+
+
+def run(argv: List[str]):
+    """Makes the g.vcf file smaller by merging similar sequential records
+    The merge is done in the following way:
+    We merge sequential records as long as there GQ value of all the records in this merge is not different
+    from each other by more than 10
```
Comment (ilyasoifer) [inline ugvc/joint/compress_gvcf.py]:
```suggestion
    from each other by more than --merge_gq_threshold
```

File: ugvc/joint/compress_gvcf.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+
+
+def get_parser() -> argparse.ArgumentParser:
+    ap_var = argparse.ArgumentParser(prog="compress_gvcf.py", description="Compress GVCF file by merging similar rows")
+    ap_var.add_argument("--input_path", help="Input gvcf file path", required=True, type=str)
+    ap_var.add_argument("--output_path", help="Output gvcf file path", required=True, type=str)
+    ap_var.add_argument("--refcall_gq_threshold",
+                        help="Keep RefCall records with GQ<refcall_threshold and not merge them",
+                        required=False,
```
Comment (ilyasoifer) [inline ugvc/joint/compress_gvcf.py:70]: Can you remind how 22 came to be?

File: ugvc/joint/compress_gvcf.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
+import pysam
```
Comment (ilyasoifer) [inline ugvc/joint/compress_gvcf.py:19]: Can you please add documentation for the script + type annotations

File: ugvc/joint/compress_gvcf.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+import argparse
+from typing import List
+import sys
+
+def get_sample_value(record, key, sample_name):
```
Comment (ilyasoifer) [inline ugvc/joint/compress_gvcf.py]: How about record.samples[sample_name].get(key,None)?

File: ugvc/joint/compress_gvcf.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    We merge sequential records as long as there GQ value of all the records in this merge is not different
+    from each other by more than 10
+    In addition, we want to keep PASS records and RefCall records with GQ<refcall_threshold and not merge them.
+    When merging we create a new record with the following parameters:
+    - chrom, pos, ref: are of the first record in this group of merged records
+    - id: '.'
+    - alts: ["<*>"]
+    - qual: 0
+    - GT: (0, 0)
+    - GQ: is the avg GQ of all the merged records
```
Comment (ilyasoifer) [inline ugvc/joint/compress_gvcf.py]: I think that min(GQ) should be taken

File: ugvc/joint/compress_gvcf.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+                        type=int)
+    return ap_var
+
+
+def run(argv: List[str]):
+    """Makes the g.vcf file smaller by merging similar sequential records
+    The merge is done in the following way:
+    We merge sequential records as long as there GQ value of all the records in this merge is not different
+    from each other by more than 10
+    In addition, we want to keep PASS records and RefCall records with GQ<refcall_threshold and not merge them.
```
Comment (ilyasoifer) [inline ugvc/joint/compress_gvcf.py]:
```suggestion
    In addition, we want to keep all PASS records and RefCall records with GQ<refcall_threshold and not merge them.
```

**Review by ilyasoifer — COMMENTED** [review]
Looks good, some comments

File: ugvc/joint/compress_gvcf.py (hunk 5 above)
Comment (mayalevy) [reply ugvc/joint/compress_gvcf.py:70]: We just looked together at some file and this is the number that you chose

File: ugvc/joint/compress_gvcf.py (hunk 6 above)
Comment (mayalevy) [reply ugvc/joint/compress_gvcf.py:19]: Can you explain what does miss in the documentation?

Comment (ilyasoifer) [reply ugvc/joint/compress_gvcf.py:19]: Just one-liner explanation of what the script does please?

**Review by ilyasoifer — APPROVED** [review]

**Pull request merged**